Script to relocate code/data/bss sections to different memory types.

varun-sha opened the pull request:
This script will relocate .text .data and .bss sections from
    required files and places it in the required memory region. This
    memory region and file are given to this python script in the form
    of a string. This script is always invoked from inside cmake.

    The main goal of this script is to provide a robust way to re-order
    the memory contents without actually having to modify the code.
    In simple terms this script will do the job of
    __attribute__((section("name"))) for a bunch of files together.

    This patch also creates a rule in the cmake to trigger the generation
    of linker_relocate.ld and code_relocation.c files.
    The linker_relocate.ld will create appropriate sections and will
    link the required functions or variables from all the selected
    files.
    The code_relocation.c will have code that is needed for
    initializing data sections and copy of text sections(if XIP).
    Also this will contain code that is needed for zeroing of bss.

    The procedure to invoke this feature is:
    1. Enable CONFIG_CODE_RELOCATION in the prj.conf
    2. Inside the CMakeLists.txt the following needs to be added
       immediately after cmake_minimum_required()

       include($ENV{ZEPHYR_BASE}/cmake/section_relocation.cmake
               NO_POLICY_SCOPE)

    3. Before the inclusion of boilerplate.cmake we need to mention
       all the files that need to get relocated.

       zephyr_kernel_code_relocate(src/*.c SRAM2)

       Where the first argument is the file/files and the second
       argument is the memory where it has be placed.
       NOTE: The file argument supports limited regular expressions.

    NOTE: Step 2 is needed only once. But step 3 can be done as many
    times as required. But step 3 has to be performed before the
    inclusion of boilerplate.cmake.

SebastianBoe reviewed (dismissed):
> Script to relocate code/data/bss sections for particular library/app
> to another memory(ccm and aon etc) or to other memory partition.

Code should not be generated unless there is no reasonable way
to write code in it's source langauge.

Could you elaborate on our options for satisfying the same use-case by
writing linker script code?

I would like to see that it was attempted, and then deemed unfeasible to write code
before one decided to generate code.

varun-sha commented:
This are my initial commits with basic testing done on qemu_cortex_m3 & relocation checked with readelf, zephyr.map/lst files.

To test script, i have created generic sram2 partition(just for test purpose), pls check below commit:-
https://github.com/varun-sha/zephyr/commit/b61f3bdff5891170a56870a2156e8b6c51a95411

to relocate particular file to particular memory or partition , add below property in CMakeLists.txt
set_property(GLOBAL PROPERTY CODE_DATA_RELOCATION_STRINGS "SRAM2:src/main.c")

varun-sha commented:
@SebastianBoe 
>Code should not be generated unless there is no reasonable way
>to write code in it's source langauge.

>Could you elaborate on our options for satisfying the same use-case by
>writing linker script code?

> would like to see that it was attempted, and then deemed unfeasible to write code
>before one decided to generate code.

[Varun]=> Function/variables can be placed into specfic memory using source code by http://www.keil.com/support/man/docs/armlink/armlink_pge1362066000009.htm
But this approach will clutter code with ifdef's depending on which memory it should go & is not generic if we want to support multiple memory types/partition's.

So i have created scripts/gen_relocate_app.py which will create another ld file called linker_relocate.ld, which has all sections which need to be relocated, extracted from obj files &  will be included in main linker.cmd. It will be enabled/disabled based on property values set in CMakeLists.txt..
All of this is done after compilation and before linking.

SebastianBoe commented:
I see this is marked as RFC, what more work needs to be done before it is in a merge-able state?

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -702,6 +702,10 @@ if(CONFIG_APPLICATION_MEMORY)
   list(APPEND LINKER_SCRIPT_DEFINES ${def})
 endif() # CONFIG_APPLICATION_MEMORY
 
+if(CONFIG_ARM)
```
How is this specific to ARM?

varun-sha replied:
Not specific to ARM, as of now it is being tested for ARM only...since this feature more time to test, initial implementation is being done and tested for ARM only, later other arch will be also be supported.

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -837,6 +842,30 @@ if(CONFIG_GEN_ISR_TABLES)
+ if(NOT ${CODE_DATA_RELOCATION_STRINGS_LOCAL} STREQUAL '')
+   set(MEM_RELOCATAION_LD "${PROJECT_BINARY_DIR}/include/generated/linker_relocate.ld")
+   message(CODE-RELOCATION LD: ${MEM_RELOCATAION_LD})
+         add_custom_command(
```
The intendation is off.

varun-sha replied:
corrected

SebastianBoe commented on cmake/section_relocation.cmake:
```diff
@@ -0,0 +1,11 @@
+   CODE_DATA_RELOCATION_STRINGS)
+  if(NOT ${CODE_DATA_RELOCATION_STRINGS_LOCAL} STREQUAL '')
+    set_property(GLOBAL APPEND_STRING PROPERTY CODE_DATA_RELOCATION_STRINGS ",")
+    set_property(GLOBAL APPEND_STRING PROPERTY CODE_DATA_RELOCATION_STRINGS ${location}:${file})
```
Can you do this in one line?

set_property(GLOBAL APPEND_STRING PROPERTY CODE_DATA_RELOCATION_STRINGS ,${location}:${file})

varun-sha replied:
done, pls check

SebastianBoe commented on cmake/section_relocation.cmake:
```diff
@@ -0,0 +1,11 @@
+function(zephyr_kernel_code_relocate file location)
+  get_property(CODE_DATA_RELOCATION_STRINGS_LOCAL GLOBAL PROPERTY
+   CODE_DATA_RELOCATION_STRINGS)
+  if(NOT ${CODE_DATA_RELOCATION_STRINGS_LOCAL} STREQUAL '')
```
It is much simpler if you maintain this list as a list instead of a string, by using APPEND instread of APPEND_STRING.

And then do a string replace just before you need to convert to a comma separated string.

varun-sha replied:
done, pls check

SebastianBoe commented on scripts/gen_relocate_app.py:
```diff
@@ -0,0 +1,172 @@
+from elftools.elf.elffile import ELFFile
+
+
+# This script will create linker comands for text. data, bss section relocation
```
What about rodata?

SebastianBoe commented on scripts/gen_relocate_app.py:
```diff
@@ -0,0 +1,172 @@
+    print("searchpath : ",searchpath)
+    if input_string == '':
+        print("Linker Relocatable script: no work to do")
+        return
```
Shouldn't this be an error?

varun-sha replied:
done, pls check

SebastianBoe commented:
> But this approach will clutter code with ifdef's depending on which memory it should go & is not generic if we want to support multiple memory types/partition's.

I see that this is limited in scope to the 'app' libraries object files.

Note that for the 'app' library one can use custom-rwdata.ld etc., see the Kconfig options in zephyr/misc/Kconfig.

This is not affected by "cluttering the code with ifdef's" as far as I can tell. Unless we can prove that there are problems with using the existing mechanism we should not introduce a code generator that covers an already covered use-case.

varun-sha commented on scripts/gen_relocate_app.py:
```diff
@@ -0,0 +1,172 @@
+            else:
+                rel_dict[split_list[0]] = [split_list[1]]
+        else:
+            print("Linker Relocatable script: wrong format in Cmakelist.txt")
```
done, pls check

varun-sha commented on scripts/gen_relocate_app.py:
```diff
@@ -0,0 +1,172 @@
+    try:
+        raw = open(linker_file, "w")
+    except IOError:
+            print("Error truncating file")
```
done, pls check

varun-sha commented:
> > But this approach will clutter code with ifdef's depending on which memory it should go & is not generic if we want to support multiple memory types/partition's.
> 
> I see that this is limited in scope to the 'app' libraries object files.
> 
> Note that for the 'app' library one can use custom-rwdata.ld etc., see the Kconfig options in zephyr/misc/Kconfig.
> 
> This is not affected by "cluttering the code with ifdef's" as far as I can tell. Unless we can prove that there are problems with using the existing mechanism we should not introduce a code generator that covers an already covered use-case.

Its not specific to "app" only, kernel files can be be relocated if zephyr_kernel_code_relocate(file mem_type) is called from kernel specific CMakeList.txt

SebastianBoe commented:
> Its not specific to "app" only, kernel files can be be relocated if zephyr_kernel_code_relocate(file mem_type) is called from kernel specific CMakeList.txt

My mistake, I misunderstood what APPLICATION_BINARY_DIR was.

SebastianBoe commented on scripts/gen_relocate_app.py:
```diff
@@ -0,0 +1,179 @@
+    for dirpath, dirs, files in os.walk(searchpath):
+        for filename in files:
+            fullname = os.path.join(dirpath,filename)
+            if re.match(".*"+obj_filename+"\.obj$",fullname):
```
What if you have two source files named util.c ?

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -837,6 +842,31 @@ if(CONFIG_GEN_ISR_TABLES)
+add_custom_target(
+  ${CODE_DATA_RELOCATION}
+  DEPENDS app
+  kernel
```
Can this break if a library other than 'app' or 'kernel' changes?

AdithyaBaglody replied:
can you please elaborate?

SebastianBoe replied:
There is no dependency from code_data_relocation to, for instance, subsys__bluetooth__controller.

So if a bluetooth file is modified, and this bluetooth file should be relocated by CODE_DATA_RELOCATION, then you will not re-generate CODE_DATA_RELOCATION, and you end up with a broken build.

You can test this by building. And then touching some source file that is not a part of 'app' or 'kernel', but that should be relocated. And then see if re-location is re-executed when "ninja/make" is invoked again (whithout cleaning the build directory).

AdithyaBaglody replied:
I understand. But how do i proceed?
Half the rules in our cmake are the same for generated files. 

We have been doing a fresh build since a long time if we expect the generated files to change.

IMHO I believe this comment is not pressing enough to stop the merge.

SebastianBoe replied:
It's true that there are some existing bugs related to this.

But nearly all C files can be modified without breaking the build.

This is not OK to merge.

AdithyaBaglody replied:
@SebastianBoe Done. please re-review.

SebastianBoe commented:
I am struggling to visualize how this will be used.

Could you show an example of a use-case where it is useful, and demonstrate how without this script this same use-case would be more difficult to achieve?

AdithyaBaglody committed:
cmake: Added rule and helper functions for code relocation.

This patch creates a rule in the cmake to trigger the generation
of linker_relocate.ld and code_relocation.c files.
The linker_relocate.ld will create appropriate sections and will
link the required functions or variables from all the selected
files.
The code_relocation.c will have code that is needed for
initializing data sections and copy of text sections(if XIP).
Also this will contain code that is needed for zeroing of bss.

The procedure to invoke this feature is:
1. Enable CONFIG_CODE_RELOCATION in the prj.conf

2. Inside CMakeList.txt in the project we need to mention
   all the files that needs to get relocated.

   zephyr_kernel_code_relocate(src/*.c SRAM2)

   Where the first argument is the file/files and the second
   argument is the memory where it has be placed.
   NOTE: The file argument supports glob expressions.

NOTE: Step 2 can be done as many times as required. And relative
paths can be given here.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
kernel: init.c: Added required hooks for the relocation

This patch splits the text section into 2 parts. The first section
will have some info regarding vector tables and debug info. The
second section will have the complete text section.
This is needed to force the required functions and data variables
the correct locations.
This is due to the behavior of the linker. The linker will only link
once and hence this text section had to be split to make room
for the generated linker script.

Added a new Kconfig CODE_DATA_RELOCATION which when enabled will
invoke the script, which does the required relocation.

Added hooks inside init.c for bss zeroing and data copy operations.
Needed when we have to copy data from ROM to required memory type.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
arch: Kconfig: Increased the text area for kobject and priv stack

When code relocation feature with userspace mode is turned on we
need a bit more memory to fit the text section for these 2
generated files.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody commented:
@SebastianBoe  Please have a look at the new patches. The architecture of this RFC has change significantly.

varun-sha commented:
> I am struggling to visualize how this will be used.
> 
> Could you show an example of a use-case where it is useful, and demonstrate how without this script this same use-case would be more difficult to achieve?

@SebastianBoe  check this sample https://github.com/zephyrproject-rtos/zephyr/pull/10892/commits/93d80184c7931727ffc19a45120f3cfdf1ffd788

SebastianBoe commented:
> This script is always invoked from inside cmake

It's actually invoked by the generated build system, not by CMake directly.

SebastianBoe commented:
> Inside the CMakeLists.txt the following needs to be added
> immediately after cmake_minimum_required().
> 
> include($ENV{ZEPHYR_BASE}/cmake/section_relocation.cmake
>            NO_POLICY_SCOPE)


No other feature has this requirement. We should remove this limitation. I believe this is possible because we are using add_custom_command and properties to store and process the data.

SebastianBoe commented:
> limited regular expressions.

This looks like a glob expression and not a subset of regex.

SebastianBoe commented:
The similarities between this PR and the __ramfunc PR should be
investigated in case there is an opportunity for re-use.

E.g. if all we really need is to run code from ram then the __ramfunc functionality should cover it. If we also need to specify which ram region, then we should try to chose one of the two mechanisms.

It's important that we use few concepts to solve many problems. And that we avoid a new mechanism for every problem.

SebastianBoe commented:
> zephyr_kernel_code_relocate

Should be "zephyr_code_relocate" to communicate that we relocate
not only kernelspace code, but also userspace code.

nashif commented:
The text you provide in the PR is fine for explaining what that is, but I want to see this documented in details with additional examples and usages etc as part of the Zephyr documentation.

varun-sha commented:
> > @SebastianBoe check this sample [93d8018](https://github.com/zephyrproject-rtos/zephyr/commit/93d80184c7931727ffc19a45120f3cfdf1ffd788)
> 
> I would like to test it. Which BOARD is it compatible with @varun-sha ?
> EDIT: Nevermind, got it to build.

@SebastianBoe : you can test on sam e70

AdithyaBaglody commented:
@SebastianBoe 
> It's not clear which code we are talking about, the C code, the
linker script, or both?

I was referring to both. No need to provide attributes and no need to create new sections.

>No other feature has this requirement. We should remove this limitation. I believe this is possible because we are using add_custom_command and properties to store and process the data.

Please help me in by providing an example. Because I need all the filenames and locations where it should go before the main CMakeLists.txt is processed. This is the only way I could think of.

> __ramfunc 

This is not quite flexible. can work for one memory location and need to added it for each function. I just looked at the code. Can you please share the PR number so that I am not missing anything here.

>This looks like a glob expression and not a subset of regex.

I was unware of the difference. Will update the same.

varun-sha committed:
doc: code_data_relocation: Details about code data relocation feature

Added basic working details of code/data/bss relocation feature,
how to use with examples & code sample details

Signed-off-by: Varun Sharma <varun.sharma@intel.com>

varun-sha commented:
> > zephyr_kernel_code_relocate
> 
> Should be "zephyr_code_relocate" to communicate that we relocate
> not only kernelspace code, but also userspace code.

Done

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,99 @@
+hence this text section had to be split to make room for the generated linker script.
+
+The ``code_relocation.c`` will have code that is needed for
+initializing data sections and copy of text sections(if XIP).
```
```suggestion
initializing data sections and a copy of text sections (if XIP).
```

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,99 @@
+
+The ``code_relocation.c`` will have code that is needed for
+initializing data sections and copy of text sections(if XIP).
+Also this will contain code that is needed for bss zeroing and
```
```suggestion
Also this contains code needed for bss zeroing and
```

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,99 @@
+
+**The procedure to invoke this feature is:**
+
+* Enable CONFIG_CODE_RELOCATION in the ``prj.conf``
```
```suggestion
* Enable :option:`CONFIG_CODE_RELOCATION` in the ``prj.conf`` file
```

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,99 @@
+**The procedure to invoke this feature is:**
+
+* Enable CONFIG_CODE_RELOCATION in the ``prj.conf``
+* Inside the ``CMakeLists.txt`` the following needs to be added immediately
```
```suggestion
* Inside the ``CMakeLists.txt`` file, add the following line immediately
```

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,99 @@
+  all the files that need to get relocated.
+
+  ``zephyr_code_relocate(src/*.c SRAM2)``
+  Where the first argument is the file/files and the second
```
add a blank line before this to start a new paragraph

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,99 @@
+
+  ``zephyr_code_relocate(src/*.c SRAM2)``
+  Where the first argument is the file/files and the second
+  argument is the memory where it has be placed.
```
```suggestion
  argument is the memory where it must be placed.
```

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,99 @@
+  Where the first argument is the file/files and the second
+  argument is the memory where it has be placed.
+
+  Note: The file argument supports limited regular expressions.
```
use a note directive (indented the same as the body text of the bullet list):
```
  .. note::  The file argument supports limited regular expressions.
     Function ``zephyr_code_relocate()`` can be called as many times as required.
     This step has to be performed before the inclusion of boilerplate.cmake.
```

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,99 @@
+  This step has to be performed before the inclusion of boilerplate.cmake.
+
+
+Additional Configuration's:-
```
```suggestion
Additional Configurations
```

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,99 @@
+
+Additional Configuration's:-
+============================
+This section shows few Configuration options that can be set in ``CMakeLists.txt``
```
```suggestion
This section shows additional configuration options that can be set in ``CMakeLists.txt``
```

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,99 @@
+============================
+This section shows few Configuration options that can be set in ``CMakeLists.txt``
+
+* if the memory is like SRAM1/SRAM2/CCD/AON then place full object in the sections
```
```suggestion
* if the memory is SRAM1, SRAM2, CCD, or AON, then place the full object in the sections
  for example:
```

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,99 @@
+
+.. code-block:: none
+
+   e.g: zephyr_code_relocate(src/file1.c SRAM2)
```
```suggestion
   zephyr_code_relocate(src/file1.c SRAM2)
```

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,99 @@
+.. code-block:: none
+
+   e.g: zephyr_code_relocate(src/file1.c SRAM2)
+        zephyr_code_relocate(src/file2.c.c SRAM)
```
indent this line the same as the previous one

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,99 @@
+   e.g: zephyr_code_relocate(src/file1.c SRAM2)
+        zephyr_code_relocate(src/file2.c.c SRAM)
+
+* if the memory type is appended with _DATA / _TEXT/ _BSS only the
```
```suggestion
* if the memory type is appended with _DATA, _TEXT, or _BSS, only the
```

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,99 @@
+        zephyr_code_relocate(src/file2.c.c SRAM)
+
+* if the memory type is appended with _DATA / _TEXT/ _BSS only the
+  selected memory is placed in the required memory region. Others are ignored.
```
```suggestion
  selected memory is placed in the required memory region.
  Others are ignored.  For example:
```

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,99 @@
+
+.. code-block:: none
+
+   e.g: zephyr_code_relocate(src/file1.c SRAM2_DATA)
```
```suggestion
  zephyr_code_relocate(src/file1.c SRAM2_DATA)
```

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,99 @@
+.. code-block:: none
+
+   e.g: zephyr_code_relocate(src/file1.c SRAM2_DATA)
+        zephyr_code_relocate(src/file2.c.c SRAM2_TEXT)
```
indent the same as the previous line

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,99 @@
+   e.g: zephyr_code_relocate(src/file1.c SRAM2_DATA)
+        zephyr_code_relocate(src/file2.c.c SRAM2_TEXT)
+
+* Multiple regions can also be appended together like SRAM2_DATA_BSS
```
```suggestion
* Multiple regions can also be appended together such as: SRAM2_DATA_BSS.
```

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,99 @@
+        zephyr_code_relocate(src/file2.c.c SRAM2_TEXT)
+
+* Multiple regions can also be appended together like SRAM2_DATA_BSS
+  this will place data and bss inside SRAM2
```
```suggestion
  This will place data and bss inside SRAM2.
```

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,99 @@
+* Multiple regions can also be appended together like SRAM2_DATA_BSS
+  this will place data and bss inside SRAM2
+
+Sample:-
```
```suggestion
Sample
```

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,99 @@
+
+Sample:-
+========
+An sample showcasing this feature is provided at
```
```suggestion
A sample showcasing this feature is provided at
```

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,99 @@
+An sample showcasing this feature is provided at
+``$ZEPHYR_BASE/samples/test_relocation/``
+
+Above sample provides an example for using the code relocation feature.
```
```suggestion
This is an example of using the code relocation feature.
```

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,99 @@
+``$ZEPHYR_BASE/samples/test_relocation/``
+
+Above sample provides an example for using the code relocation feature.
+This example will place text,data,bss from 3 files to various parts in the SRAM.
```
```suggestion
This example will place text, data, and bss from 3 files to various parts in SRAM
using a custom linker file derived from ``include/arch/arm/cortex_m/scripts/linker.ld``.
```

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,99 @@
+
+Above sample provides an example for using the code relocation feature.
+This example will place text,data,bss from 3 files to various parts in the SRAM.
+For this a custom linker file is used
```
... and delete these two lines

varun-sha commented:
@dbkinder : pls check, i have updated the documentation with resolution's

varun-sha commented:
> Found the following issues, please fix and resubmit:
> 
> ## License/Copyright issues
> * scripts/gen_relocate_app.py missing license.
> * scripts/gen_relocate_app.py missing copyright.

Resolved

varun-sha commented:
@SebastianBoe @dbkinder @nashif : have resolved review comments , please check.

SebastianBoe commented on scripts/gen_relocate_app.py:
```diff
@@ -0,0 +1,352 @@
+        for section in sections:
+
+            if (".text." in  section.name) or (".rodata." in  section.name):
+                full_list_of_sections["text"].append(section.name)
```
Why is rodata placed in the text section? That is unexpected.

SebastianBoe commented:
> have resolved review comments , please check.

Not all comments have been addressed. e.g. https://github.com/zephyrproject-rtos/zephyr/pull/10892#discussion_r231806038

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -699,6 +699,13 @@ if(CONFIG_APPLICATION_MEMORY)
+  set(CODE_RELOCATION_DEP code_relocation_source_lib)
+  # This is needed to force the .common type to .bss type
+  # Without which the common variables cant be placed in the required section
+  zephyr_compile_options(-fno-common)
```
Why can't the common section be relocated just like .bss is?

Could you elaborate on what is special about it in this comment?

SebastianBoe replied:
> Common symbols only exist in object files, not in executable files. The idea of COMMON symbols in Unix is to allow multiple external definitions of a same variable (in different compilation units) under a single common symbol under certain conditions.

Ah, I see, so it won't exist in the executable. I wonder how this will affect variables declared with extern.

AdithyaBaglody replied:
>I think we should apply no-common either everywhere

That is quite outside the scope of this PR. Can do it as a sequel to this PR.

>If there exist two non-extern declarations of the same variable 

This is a feature of the GCC itself and the only reason i am restricting it is because this PR doesn't work without it.

SebastianBoe replied:
> That is quite outside the scope of this PR.

I'm not sure it is.

Imagine that in dozens of places, we are relying on the existence of common symbols.

This would mean that often, when a user enables CONFIG_CODE_RELOCATION, he unexpectedly gets a linker error. Essentially, CONFIG_CODE_RELOCATION is not well-integrated with the rest of Zephyr. This would make this feature unstable, and therefore not
something we want to present to users.

But we don't know if this is the case. Perhaps common symbols are only relied upon in a handful of situations. And these situations can easily be ported to not rely on common symbols. If so, it is best if we port all of these cases by always applying no-common. So that users do not have to discover and report them for us.

In either case. We don't want to apply no-common only when CONFIG_CODE_RELOCATION is enabled.

Essentially, CODE_RELOCATION requires that the C language behaves in a certain way, and for consistency, and maintainability reasons, we want to have it always behave the same.

AdithyaBaglody replied:
This is a breaking change. What if someone is relying on this feature outside the zephyr tree?. 
If someone is enabling this code relocation feature, he knows what he is getting himself into. But enabling globally by default, I am skeptical.
@nashif Can you please say if it is okay to enable this flag globally(-fno-common)?

SebastianBoe replied:
> What if someone is relying on this feature outside the zephyr tree?

If someone relies on it, and needs to continue to rely on it. Then we should find a way to avoid enabling -fno-common, or else CODE_RELOCATION should not be in-tree.

I am against having the semantics of COMMON be different depending on what features are enabled. The language must always behave the same.

AdithyaBaglody replied:
Enabled it globally.

AdithyaBaglody replied:
@SebastianBoe please re-review

SebastianBoe replied:
Done. https://github.com/zephyrproject-rtos/zephyr/pull/10892#discussion_r236717462

SebastianBoe commented on samples/test_relocation/prj.conf:
```diff
@@ -0,0 +1,4 @@
+CONFIG_CODE_DATA_RELOCATION=y
+CONFIG_HAVE_CUSTOM_LINKER_SCRIPT=y
+CONFIG_CUSTOM_LINKER_SCRIPT="linker_arm_sram2.ld"
```
We must have a sample that does not require copying the cortex linker script.

The use-case of this functionality is that one doesn't need to maintain a lot of linker script code, but if we need to maintain an entire copy of the linker script, then we are not achieving this.

AdithyaBaglody replied:
the only reason this is needed is because we are creating new memory regions to showcase this feature. if someone is using this feature they will have their own linker script in SOC anyway.

This unfortunately is a must to showcase an example.

SebastianBoe replied:
Couldn't we use some platform that already has multiple regions?

Then we wouldn't need to create new ones.

IMHO the cost of maintaining another copy of the cortex linkerscript is greater than the benefit we get from the test coverage of CONFIG_CODE_DATA_RELOCATION.

AdithyaBaglody replied:
>IMHO the cost of maintaining another copy of the cortex linkerscript is greater than the benefit we get from the test coverage of CONFIG_CODE_DATA_RELOCATION.

@SebastianBoe  Makes sense. I found a workaround. Please re-review.

SebastianBoe replied:
This is OK. Resolving this point.

AdithyaBaglody committed:
samples: code_relocation: An example for code relocation feature.

This sample provides an example for using the code relocation
feature. This example will place text,data,bss from 3 files to
various parts in the SRAM. For this a custom linker file is used
which is derived from include/arch/arm/cortex_m/scripts/linker.ld.

Signed-off-by: Varun Sharma <varun.sharma@intel.com>
Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
scripts: gen_relocate_app.py: Script for relocating files in memory.

This script will relocate .text .data and .bss sections from
required files and places it in the required memory region. This
memory region and file are given to this python script in the form
of a string. Refer to the script for the format of this string and
the procedure to invoke it.

The main goal of this script is to provide a robust way to re-order
the memory contents without actually having to modify the code
(C source code and the linker code).
In simple terms this script will do the job of
__attribute__((section("name"))) for a bunch of files together.

Signed-off-by: Varun Sharma <varun.sharma@intel.com>
Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
tests: benchmarks: timing_info: Remove common variables

This patch ensures that there are no .common variables in the
test case.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
drivers: ipm: ipm_quark_se: Remove object declaration from header

The header ipm_quark_se.h was creating a object. Hence removed it
and placed the same in ipm_quark_se.c

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
include: drivers: pci: pci_mgr.h: Create a typdef instead of obj.

This header was creating an object instead of a typdef. Fixed the
bug.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
include: stats.h: Packed attribute was incorrect.

This header was actually creating a struct in the bss. Fixed it.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
soc: ti_simplelink: cc32xx: soc.h: Incorrect enum definition.

Instead of creating an typedef enum it was creating an object.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
tests: subsys: fs: Fixed headers which were creating objects.

These tests were creating objects from header file in the bss.
Fixed it by moving the objects to appropriate object files.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
samples: rpl_border_router: Fixed headers files.

This sample was creating objects from header file in the bss.
Fixed it by moving the objects to appropriate object files.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
CMakeLists.txt: Enable -fno-common globally.

This is needed to force the .common type to .bss type.
This will force the .common variables to go into bss section
in the compiled object file.
As of now the movement to the .bss section was happening only
during the final linking stage.

The -fno-common option specifies that the compiler should place
uninitialized global variables in the data section of the object
file, rather than generating them as common blocks.  This has the
effect that if the same variable is declared (without "extern")
in two different compilations, you will get a multiple-definition
error when you link them.

e.g:
// file a.c
// file-scope

int b;

// file b.c
// file-scope

int b;

If there exist two non-extern declarations of the same variable
then no-common will cause a linker error.
This sequence would have compiled before, but will now cause a
linker error.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -136,6 +136,7 @@ zephyr_compile_options(
   -imacros ${AUTOCONF_H}
```
The commit message needs to be re-written.

It should be clear to the reader that this is a
breaking change and how to identify whether he is affected.

The discussion leading up to this change had a good explanation
I believe.

This is so that, when this breaks something, it is easier to identify
where it broke.

AdithyaBaglody replied:
@SebastianBoe yep make it a bit more verbose. Please  re-review.

SebastianBoe commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,102 @@
+  after cmake_minimum_required()
+  Note: This is needed only once.
+
+  ``include($ENV{ZEPHYR_BASE}/cmake/section_relocation.cmake NO_POLICY_SCOPE)``
```
Requiring that the user does this is not aligned with how Zephyr works.

This is how we can fix this:

Move "function(zephyr_code_relocate file location)" into boilerplate.cmake.

Modify 'zephyr_code_relocate' to work on a custom target's
COMPILE_DEFINITION property instead of the global property
'CODE_DATA_RELOCATION_STRINGS'. See how the COMPILE_DEFINITION
property of 'memory_space_kernel'[0] is used for reference. They
are solving the same problem.  [0]
https://github.com/zephyrproject-rtos/zephyr/pull/8451/files#diff-af3b638bc2a3e6c650974192a53c7291R592

Don't de-reference 'CODE_DATA_RELOCATION_STRINGS' with
'get_property' in zephyr/CMakeLists.txt. But instead use a
generator expression to defer evaluation until generation
time. See [0] for how to do this.

Update the documentation accordingly.

AdithyaBaglody replied:
@SebastianBoe Done. Please re-review.

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,97 @@
+*******
+The memory region and file are given to this python script in the form of a string.
+
+Example of such a string would be:
```
```suggestion
An example of such a string is:
```

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,97 @@
+Example of such a string would be:
+``SRAM2:/home/xyz/zephyr/samples/hello_world/src/main.c,SRAM1:/home/xyz/zephyr/samples/hello_world/src/main2.c``
+
+This script is invoked with following parameters:
```
```suggestion
This script is invoked with the following parameters:
```

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,97 @@
+
+This script also trigger the generation of ``linker_relocate.ld`` and ``code_relocation.c`` files.
+The ``linker_relocate.ld`` file creates appropriate sections and
+link the required functions or variables from all the selected files.
```
```suggestion
links the required functions or variables from all the selected files.
```

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,97 @@
+The ``linker_relocate.ld`` file creates appropriate sections and
+link the required functions or variables from all the selected files.
+
+**Note**:
```
This would present better as:
```
.. note:: After running this script, the text section is split into 2 parts.
   The first section will have some info regarding vector tables and debug info.
   The second section will have the complete text section.
   This is needed to force the required functions and data variables to the correct locations.
   This is due to the behavior of the linker. The linker will only link once and
   hence this text section had to be split to make room for the generated linker script.
```

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,97 @@
+hence this text section had to be split to make room for the generated linker script.
+
+The ``code_relocation.c`` file has code that is needed for
+initializing data sections and a copy of text sections (if XIP).
```
```suggestion
initializing data sections, and a copy of the text sections (if XIP).
```

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,97 @@
+
+* Enable :option:`CONFIG_CODE_DATA_RELOCATION` in the ``prj.conf`` file
+
+* Inside the ``CMakeLists.txt`` file in the project. we need to mention
```
```suggestion
* Inside the ``CMakeLists.txt`` file in the project, mention
```

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,97 @@
+* Enable :option:`CONFIG_CODE_DATA_RELOCATION` in the ``prj.conf`` file
+
+* Inside the ``CMakeLists.txt`` file in the project. we need to mention
+  all the files that needs to get relocated.
```
```suggestion
  all the files that need relocation.
```

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,97 @@
+* if the memory is SRAM1, SRAM2, CCD, or AON, then place the full object in the sections
+  for example:
+
+.. code-block:: none
```
This code block (and contents) should be indented by two spaces to be treated as part of the bullet list item.

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,97 @@
+  selected memory is placed in the required memory region.
+  for example:
+
+.. code-block:: none
```
same here, indent the code-block directive (and contents) by two more spaces

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,97 @@
+``$ZEPHYR_BASE/samples/test_relocation/``
+
+This is an example of using the code relocation feature.
+This example will place .text, .data, .bss from 3 files to various parts in the SRAM.
```
```suggestion
This example will place .text, .data, .bss from 3 files to various parts in the SRAM
```

dbkinder commented on doc/index.rst:
```diff
@@ -57,6 +57,7 @@ Source code for the Zephyr Project is maintained in the Zephyr Project's
    subsystems/subsystems.rst
    contribute/contribute.rst
    security/security.rst
+   code_data_relocation/relocation.rst
```
This doesn't seem to be the right place to put this document in the overall site organization.  I'd recommend this go into the developer guide (application/index.rst) and not the top level index.  This means also moving the relocation.rst file into the application folder.

dbkinder commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,97 @@
+.. _code_data_relocation:
```
As noted elsewhere, this document should go into the developer guide area, so move this from ``doc/code_data_relocation/relocation.rst`` to ``doc/application/relocation.rst`` and edit the ``doc/application/index.rst`` to include ``relocation.rst`` into the toc there.

dbkinder reviewed (changes requested):
This document is better placed in the Developer Guide area rather than in the top-level toc.

AdithyaBaglody commented on doc/code_data_relocation/relocation.rst:
```diff
@@ -0,0 +1,97 @@
+
+Overview
+********
+This script will relocate .text, .rodata, .data and .bss sections from
```
@dbkinder I don't understand why there needs to a comma before an "and" in this case. Hence I haven't fixed this.
Others are done. Please re-review.

dbkinder replied:
With so many contributors, we've tried to standardize on some simple editorial guidelines such as using English spellings (vs. British) and using the [Microsoft Writing Style Guide](https://docs.microsoft.com/en-us/style-guide/welcome/), and [Merriam-Webster Dictionary](http://www.merriam-webster.com/) as guidelines.  Use of the Oxford (or serial) comma is the recommendation, to avoid potential confusion with lists.

AdithyaBaglody replied:
ohh okay. Changed to the recommendation.
@dbkinder Please re-review.

dbkinder reviewed (approved):
LGTM, thanks.

nashif commented on kernel/Kconfig:
```diff
@@ -279,6 +279,13 @@ config WAITQ_DUMB
 
 endchoice # WAITQ_ALGORITHM
 
+config CODE_DATA_RELOCATION
```
why is this under kernel/Kconfig? Please move out of here to Kconfig.zephyr under linker options for example.

nashif commented on kernel/Kconfig:
```diff
@@ -279,6 +279,13 @@ config WAITQ_DUMB
+	depends on ARM
+	help
+	  When selected this will relocate .text, data and .bss sections from
+	  required files and places it in the required memory region.
```
required filed? require by what? Maybe specified file.

nashif commented on samples/test_relocation/sample.yaml:
```diff
@@ -0,0 +1,12 @@
+      regex:
+      - "Hello World! (.*)"
+    platform_whitelist: qemu_cortex_m3 mps2_an385 sam_e70_xplained
+    tags: introduction
```
tags: linker

nashif commented on samples/test_relocation/sample.yaml:
```diff
@@ -0,0 +1,12 @@
+sample:
+  description: Code data relocation Sample
+  name: test relocation
```
name: Code Relocation

nashif commented on samples/test_relocation/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+cmake_minimum_required(VERSION 3.8.2)
```
Move this sample under samples/application_development

Also, do not call it test_***, if it is a test, it should not be under samples/

nashif commented on scripts/gen_relocate_app.py:
```diff
@@ -0,0 +1,370 @@
+#!/usr/bin/env python3
```
Please spend some time on this script and fix the syntax, you get it to a higher score, at least 8.00 :)

```
************* Module gen_relocate_app
scripts/gen_relocate_app.py:12:0: C0301: Line too long (110/100) (line-too-long)
scripts/gen_relocate_app.py:75:29: C0326: Exactly one space required before assignment
extern_linker_var_declaration= """
                             ^ (bad-whitespace)
scripts/gen_relocate_app.py:101:15: C0326: Exactly one space required before assignment
memset_template= """
               ^ (bad-whitespace)
scripts/gen_relocate_app.py:108:26: C0326: No space allowed after bracket
        full_lib = ELFFile( f)
                          ^ (bad-whitespace)
scripts/gen_relocate_app.py:109:0: C0325: Unnecessary parens after 'if' keyword (superfluous-parens)
scripts/gen_relocate_app.py:110:40: C0326: Exactly one space required after comma
            print("Error parsing file: ",filename)
                                        ^ (bad-whitespace)
scripts/gen_relocate_app.py:113:19: C0326: No space allowed after bracket
        sections = [ x for x in full_lib.iter_sections()]
                   ^ (bad-whitespace)
scripts/gen_relocate_app.py:118:0: C0325: Unnecessary parens after 'if' keyword (superfluous-parens)
scripts/gen_relocate_app.py:121:0: C0325: Unnecessary parens after 'if' keyword (superfluous-parens)
scripts/gen_relocate_app.py:124:0: C0325: Unnecessary parens after 'if' keyword (superfluous-parens)
scripts/gen_relocate_app.py:127:0: C0325: Unnecessary parens after 'if' keyword (superfluous-parens)
scripts/gen_relocate_app.py:135:0: C0325: Unnecessary parens after 'if' keyword (superfluous-parens)
scripts/gen_relocate_app.py:136:26: C0326: No space allowed after bracket
                symbols = [ x for x in section.iter_symbols()]
                          ^ (bad-whitespace)
scripts/gen_relocate_app.py:139:0: C0301: Line too long (120/100) (line-too-long)
scripts/gen_relocate_app.py:150:49: C0326: Exactly one space required after comma
        memory_type = memory_type.replace("_TEXT","")
                                                 ^ (bad-whitespace)
scripts/gen_relocate_app.py:153:51: C0326: Exactly one space required after comma
        memory_type = memory_type.replace("_RODATA","")
                                                   ^ (bad-whitespace)
scripts/gen_relocate_app.py:156:49: C0326: Exactly one space required after comma
        memory_type = memory_type.replace("_DATA","")
                                                 ^ (bad-whitespace)
scripts/gen_relocate_app.py:159:48: C0326: Exactly one space required after comma
        memory_type = memory_type.replace("_BSS","")
                                                ^ (bad-whitespace)
scripts/gen_relocate_app.py:167:0: C0330: Wrong continued indentation before block (add 4 spaces).
                full_list_of_sections[iter_sec] != []):
                ^   | (bad-continuation)
scripts/gen_relocate_app.py:173:64: C0326: No space allowed before bracket
        tmp_list = {"text":[], "rodata":[], "data":[], "bss":[] }
                                                                ^ (bad-whitespace)
scripts/gen_relocate_app.py:176:0: C0330: Wrong continued indentation before block (add 4 spaces).
                full_list_of_sections[iter_sec] != []):
                ^   | (bad-continuation)
scripts/gen_relocate_app.py:197:0: C0325: Unnecessary parens after 'if' keyword (superfluous-parens)
scripts/gen_relocate_app.py:229:21: C0326: Exactly one space required after assignment
    generate_section ={"text":False, "rodata":False, "data":False, "bss":False }
                     ^ (bad-whitespace)
scripts/gen_relocate_app.py:229:79: C0326: No space allowed before bracket
    generate_section ={"text":False, "rodata":False, "data":False, "bss":False }
                                                                               ^ (bad-whitespace)
scripts/gen_relocate_app.py:233:58: C0326: Exactly one space required after comma
            memory_type = memory_type.replace(section_name,"")
                                                          ^ (bad-whitespace)
scripts/gen_relocate_app.py:244:24: C0326: Exactly one space required after comma
    for mtype in ["text","rodata", "data"]:
                        ^ (bad-whitespace)
scripts/gen_relocate_app.py:263:0: W0301: Unnecessary semicolon (unnecessary-semicolon)
scripts/gen_relocate_app.py:287:0: C0330: Wrong continued indentation (add 12 spaces).
            help="input src:memory type(sram2 or ccm or aon etc) string")
            ^           | (bad-continuation)
scripts/gen_relocate_app.py:291:67: C0326: No space allowed before keyword argument assignment
    parser.add_argument("-v", "--verbose", action="count", default =0,
                                                                   ^ (bad-whitespace)
scripts/gen_relocate_app.py:304:51: C0326: Exactly one space required after comma
                    fullname = os.path.join(dirpath,filename1)
                                                   ^ (bad-whitespace)
scripts/gen_relocate_app.py:325:46: C0326: Exactly one space required after comma
            print("Memory region ", mem_region," Selected for file:", file_name_list)
                                              ^ (bad-whitespace)
scripts/gen_relocate_app.py:331:0: C0325: Unnecessary parens after 'return' keyword (superfluous-parens)
scripts/gen_relocate_app.py:346:20: C0326: Exactly one space required after comma
    for  memory_type,files in rel_dict.items():
                    ^ (bad-whitespace)
scripts/gen_relocate_app.py:347:30: C0326: Exactly one space required after assignment
        full_list_of_sections ={"text":[], "rodata":[], "data":[], "bss":[] }
                              ^ (bad-whitespace)
scripts/gen_relocate_app.py:347:76: C0326: No space allowed before bracket
        full_list_of_sections ={"text":[], "rodata":[], "data":[], "bss":[] }
                                                                            ^ (bad-whitespace)
scripts/gen_relocate_app.py:1:0: C0111: Missing module docstring (missing-docstring)
scripts/gen_relocate_app.py:37:0: C0103: Constant name "print_template" doesn't conform to UPPER_CASE naming style (invalid-name)
scripts/gen_relocate_app.py:41:0: C0103: Constant name "group_start_end_memory_seq" doesn't conform to UPPER_CASE naming style (invalid-name)
scripts/gen_relocate_app.py:45:0: C0103: Constant name "section_load_memory_seq" doesn't conform to UPPER_CASE naming style (invalid-name)
scripts/gen_relocate_app.py:49:0: C0103: Constant name "load_address_location_flash" doesn't conform to UPPER_CASE naming style (invalid-name)
scripts/gen_relocate_app.py:50:0: C0103: Constant name "load_address_location_bss" doesn't conform to UPPER_CASE naming style (invalid-name)
scripts/gen_relocate_app.py:53:0: C0103: Constant name "linker_section_seq" doesn't conform to UPPER_CASE naming style (invalid-name)
scripts/gen_relocate_app.py:68:0: C0103: Constant name "source_code_includes" doesn't conform to UPPER_CASE naming style (invalid-name)
scripts/gen_relocate_app.py:75:0: C0103: Constant name "extern_linker_var_declaration" doesn't conform to UPPER_CASE naming style (invalid-name)
scripts/gen_relocate_app.py:82:0: C0103: Constant name "data_copy_function" doesn't conform to UPPER_CASE naming style (invalid-name)
scripts/gen_relocate_app.py:89:0: C0103: Constant name "bss_zeroing_function" doesn't conform to UPPER_CASE naming style (invalid-name)
scripts/gen_relocate_app.py:96:0: C0103: Constant name "memcpy_template" doesn't conform to UPPER_CASE naming style (invalid-name)
scripts/gen_relocate_app.py:101:0: C0103: Constant name "memset_template" doesn't conform to UPPER_CASE naming style (invalid-name)
scripts/gen_relocate_app.py:106:0: C0111: Missing function docstring (missing-docstring)
scripts/gen_relocate_app.py:107:33: C0103: Variable name "f" doesn't conform to snake_case naming style (invalid-name)
scripts/gen_relocate_app.py:111:12: E1101: Module 'os' has no 'exit' member; maybe '_exit'? (no-member)
scripts/gen_relocate_app.py:144:0: C0111: Missing function docstring (missing-docstring)
scripts/gen_relocate_app.py:184:0: C0111: Missing function docstring (missing-docstring)
scripts/gen_relocate_app.py:185:4: W0621: Redefining name 'string' from outer scope (line 29) (redefined-outer-name)
scripts/gen_relocate_app.py:190:0: C0111: Missing function docstring (missing-docstring)
scripts/gen_relocate_app.py:192:4: W0621: Redefining name 'string' from outer scope (line 29) (redefined-outer-name)
scripts/gen_relocate_app.py:214:0: C0111: Missing function docstring (missing-docstring)
scripts/gen_relocate_app.py:215:4: W0621: Redefining name 'string' from outer scope (line 29) (redefined-outer-name)
scripts/gen_relocate_app.py:223:36: C0103: Variable name "fw" doesn't conform to snake_case naming style (invalid-name)
scripts/gen_relocate_app.py:226:0: C0111: Missing function docstring (missing-docstring)
scripts/gen_relocate_app.py:258:0: C0111: Missing function docstring (missing-docstring)
scripts/gen_relocate_app.py:259:4: W0621: Redefining name 'string' from outer scope (line 29) (redefined-outer-name)
scripts/gen_relocate_app.py:275:35: C0103: Variable name "fw" doesn't conform to snake_case naming style (invalid-name)
scripts/gen_relocate_app.py:279:0: C0111: Missing function docstring (missing-docstring)
scripts/gen_relocate_app.py:280:4: C0103: Constant name "args" doesn't conform to UPPER_CASE naming style (invalid-name)
scripts/gen_relocate_app.py:280:4: W0601: Global variable 'args' undefined at the module level (global-variable-undefined)
scripts/gen_relocate_app.py:296:0: C0111: Missing function docstring (missing-docstring)
scripts/gen_relocate_app.py:296:0: R1710: Either all return statements in a function should return an expression, or none of them should. (inconsistent-return-statements)
scripts/gen_relocate_app.py:300:17: W0612: Unused variable 'dirs' (unused-variable)
scripts/gen_relocate_app.py:309:0: C0111: Missing function docstring (missing-docstring)
scripts/gen_relocate_app.py:314:8: E1101: Module 'os' has no 'exit' member; maybe '_exit'? (no-member)
scripts/gen_relocate_app.py:334:0: C0111: Missing function docstring (missing-docstring)
scripts/gen_relocate_app.py:342:4: W0612: Unused variable 'raw' (unused-variable)
scripts/gen_relocate_app.py:25:0: W0611: Unused import sys (unused-import)
scripts/gen_relocate_app.py:28:0: W0611: Unused import re (unused-import)
scripts/gen_relocate_app.py:29:0: W0611: Unused import string (unused-import)
scripts/gen_relocate_app.py:31:0: C0411: standard import "import glob" should be placed before "from elftools.elf.elffile import ELFFile" (wrong-import-order)
scripts/gen_relocate_app.py:32:0: C0411: standard import "import warnings" should be placed before "from elftools.elf.elffile import ELFFile" (wrong-import-order)

-----------------------------------
Your code has been rated at 5.35/10
```

nashif replied:
pylint

AdithyaBaglody replied:
@nashif comments handled. please re-review.

AdithyaBaglody commented:
@nashif can we merge this now.
